knodle/trainer/crossweigh_weighing/data_splitting_by_rules.py

```diff
@@ -0,0 +1,289 @@
+        other_sample_ids=other_sample_ids
+    )
+
+    train_samples, train_labels, _ = get_samples_labels_idx_by_rule_id(
+        data_features, labels, train_rules_idx, rule_id2samples_ids, check_intersections=test_idx,
+        other_sample_ids=other_sample_ids
+    )
+
+    train_dataset = input_labels_to_tensordataset(train_samples, train_labels)
+    test_dataset = input_info_labels_to_tensordataset(test_samples, test_idx, test_labels)
+
+    logger.info(
+        f"Fold {fold}     Rules in training set: {len(train_rules_idx)}, rules in test set: {len(test_rules_idx)}, "
+        f"samples in training set: {len(train_samples)}, samples in test set: {len(test_samples)}"
+    )
+
+    return train_dataset, test_dataset
+
+
+def calculate_rules_indices(rules_idx: list, fold: int, num_folds: int) -> (np.ndarray, np.ndarray):
+    """
+    Calculates the indices of the samples which are to be included in training and test sets for k-fold cross validation.
+
+    :param rules_idx: all rules indices (shuffled) that are to be splitted into cw training & cw test set rules
+    :param fold: number of a current hold-out fold
+
+    :return: two arrays containing indices of rules that will be used for cw training and cw test set accordingly
+    """
+    test_rules_idx = rules_idx[fold::num_folds]
+    train_rules_idx = [rule_id for rule_id in rules_idx if rule_id not in test_rules_idx]
+
+    if not set(test_rules_idx).isdisjoint(set(train_rules_idx)):
+        raise ValueError("Splitting into train and test rules is done incorrectly.")
+
+    return train_rules_idx, test_rules_idx
+
+
+def get_samples_labels_idx_by_rule_id(
+        data_features: np.ndarray, labels: np.ndarray, indices: list, rules_samples_ids_dict: Dict,
+        check_intersections: np.ndarray = None, other_sample_ids: list = None
+) -> (TensorDataset, np.ndarray, np.ndarray):
+    """
+    Extracts the samples and labels from the original matrices by indices. If intersection is filled with
+    another sample matrix, it also checks whether the sample is not in this other matrix yet.
+
+    :param data_features: numpy array with encoded data samples (num_samples, num_features)
+    :param labels: all training samples labels (num_samples, num_classes)
+    :param indices: indices of rules; samples, where these rules matched & their labels are to be included in set
+    :param rules_samples_ids_dict: dictionary that contains information about corresponding from rules to sample ids.
+    :param check_intersections: optional parameter that indicates that intersections should be checked (used to
+    exclude the sentences from the training set which are already in the test set)
+    :param other_sample_ids: a list of sample ids that belong to the other class. They won't be included in the test
+    set, but only to the training set.
+
+    :return: samples, labels and indices in the original matrix
+    """
+    sample_ids = [list(rules_samples_ids_dict.get(idx)) for idx in indices]
```

Comment:
Is it possible that "idx" does not exist in "rules_samples_ids_dict"? To my understanding this should not be the case. A [None] sample_ids list would probably also cause problems later, so better catch the error here, what do you think?
```suggestion
    sample_ids = [list(rules_samples_ids_dict[idx]) for idx in indices]
```